File: kotlin-orders/kotlin-orders-dto/src/main/java/com/mvv/bank/orders/rest/entities/BaseOrder.kt
package com.mvv.bank.orders.rest.entities

import java.math.BigDecimal
import java.time.ZonedDateTime


sealed class BaseOrder {
    var id: Long? = null
    lateinit var user: String
    lateinit var market: String

    lateinit var orderType: OrderType
    lateinit var orderState: OrderState
    lateinit var side: Side
    lateinit var buySellType: BuySellType
    lateinit var volume: BigDecimal

    var limitPrice: Amount? = null
    var stopPrice:  Amount? = null
    var dailyExecutionType: DailyExecutionType? = null

    var placedAt:   ZonedDateTime? = null
    var executedAt: ZonedDateTime? = null
    var canceledAt: ZonedDateTime? = null
    var expiredAt:  ZonedDateTime? = null
}

File: scala-orders/src/main/scala/com/mvv/bank/orders/rest/entities/BaseOrder.scala
package com.mvv.bank.orders.rest.entities

import com.mvv.bank.orders.rest.*

import java.time.ZonedDateTime
import scala.compiletime.uninitialized


//sealed
class BaseOrder {
    //var id: Long = uninitialized
    var id: Option[Long] = None
    var user: String = uninitialized
    var market: String = uninitialized

    var orderType: OrderType = uninitialized
    var orderState: OrderState = uninitialized
    var side: Side = uninitialized
    var buySellType: BuySellType = uninitialized
    var volume: BigDecimal = uninitialized

    var limitPrice: Option[Amount] = None
    var stopPrice:  Option[Amount] = None
    var dailyExecutionType: Option[DailyExecutionType] = None

    var resultingPrice:  Option[Amount] = None

    var placedAt:   Option[ZonedDateTime] = None
    var executedAt: Option[ZonedDateTime] = None
    var canceledAt: Option[ZonedDateTime] = None
    var expiredAt:  Option[ZonedDateTime] = None
}


// CashOrder

class CashOrder extends BaseOrder {
  var buyCurrency:   String = uninitialized
  var sellCurrency:  String = uninitialized
  var priceCurrency: String = uninitialized
  var resultingRate: Option[FxRate] = None
}
